### PR TITLE
Notify listen

### PR DESCRIPTION
**Pull request by pedantix**
Added Listen and Notify functionality.

Resolves: https://github.com/vapor/postgresql/issues/40

### CODE REVIEW

**Comment by pedantix** [comment]
so the test timed out on the select and I have no idea why locally it runs well. I don't have sufficient authority to ssh into the test container to investigate.

**Comment by pedantix** [comment]
found out why tests were failing... there may or may not be a one or two strings in the notificationResponse

**Comment by tanner0101** [comment]
@shnhrrsn could you take a look at the merge conflicts?

**Comment by pedantix** [comment]
on it

File: Sources/PostgreSQL/Connection/PostgreSQLConnection.swift [hunk 1]
```diff
@@ -134,8 +134,17 @@ public final class PostgreSQLConnection {
...
     public func close() {
-        channel.close(promise: nil)
+        if let beforeClose = beforeClose {
+            _ = beforeClose(self).then { _ in
+                self.channel.close(mode: CloseMode.all)
+            }
+        } else {
+            channel.close(promise: nil)
+        }
+
```
Comment (MrMage) [inline Sources/PostgreSQL/Connection/PostgreSQLConnection.swift]: Nit: extra line

File: Tests/PostgreSQLTests/PostgreSQLConnectionTests.swift [hunk 2]
```diff
@@ -342,6 +342,25 @@ class PostgreSQLConnectionTests: XCTestCase {
...
+        let messageText = "Bar"
+        var messageReceived = false
+        
+        try listenConn.listen(channelName) { text in
+            messageReceived = text == messageText
+        }.catch({ err in XCTFail("error \(err)") })
+
+        try notifyConn.notify(channelName, message: messageText).wait()
+
+        sleep(1) // Wait for any delay in the message being received
```
Comment (MrMage) [inline Tests/PostgreSQLTests/PostgreSQLConnectionTests.swift]: How about creating a test expectation and waiting on that instead? Should be faster and more reliable than sleeping. See e.g. https://github.com/grpc/grpc-swift/blob/d247c61b4de064b121b40829df62c69849cafd6b/Tests/SwiftGRPCTests/EchoTests.swift#L72 for an example that works under Linux.

File: Sources/PostgreSQL/Connection/PostgreSQLConnection+NotifyAndListen.swift [hunk 3]
```diff
@@ -0,0 +1,29 @@
+import Async
+
+extension PostgreSQLConnection {
+    public func listen(
+        _ channelName: String,
+        handler: @escaping (String) throws -> ()
+        ) throws -> Future<Void> {
+        beforeClose = { conn in
```
Comment (MrMage) [inline Sources/PostgreSQL/Connection/PostgreSQLConnection+NotifyAndListen.swift]: Sorry, but what would happen if we try to listen to more than one channel? Wouldn’t the first one never get unlistened then? Or is it a bad idea to use a connection that listens to one specific channel for anything else, anyway?

If calling `listen` means that this connection should no longer be used for anything else, I think that should be clearly documented.

(Technically, there should be little reason not to listen to more than one channel on the same connection (they can be switched on via process id/channel name), but I understand that that might be infeasible given the way `QueueHandler` works at the moment. Still worth considering for the future, though, given that Postgres has a somewhat low connection limit (100) by default.)

File: Sources/PostgreSQL/Connection/PostgreSQLConnection+NotifyAndListen.swift [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+            let query = PostgreSQLQuery(query: "UNLISTEN \(channelName);")
+            return conn.send([.query(query)], onResponse: { _ in })
+        }
+        let query = PostgreSQLQuery(query: "LISTEN \(channelName);")
+        return queue.enqueue([.query(query)], onInput: { message in
+            switch message {
+            case let .notificationResponse(notification):
+                try handler(notification.message)
```
Comment (MrMage) [inline Sources/PostgreSQL/Connection/PostgreSQLConnection+NotifyAndListen.swift]: Related to the above: Sorry for the stupid question, but is it guaranteed that all payloads sent to that particular channel (but not those sent to other channels) are handled by this onInput handler? I guess yes, because the handler returns `false`, but then see the above comment.

Explanation: I have no idea whether PostgreSQLConnection is „smart“ enough to automatically dispatch all subsequent messages to this channel to this onInput handler, or whether the handler would be discarded after having received its first notification.

File: Tests/PostgreSQLTests/PostgreSQLConnectionTests.swift [hunk 5]
```diff
@@ -342,6 +342,25 @@ class PostgreSQLConnectionTests: XCTestCase {
...
+        let listenConn = try PostgreSQLConnection.makeTest()
+        let channelName = "Foo"
+        let messageText = "Bar"
+        var messageReceived = false
+        
+        try listenConn.listen(channelName) { text in
+            messageReceived = text == messageText
+        }.catch({ err in XCTFail("error \(err)") })
+
+        try notifyConn.notify(channelName, message: messageText).wait()
```
Comment (MrMage) [inline Tests/PostgreSQLTests/PostgreSQLConnectionTests.swift]: I would suggest sending two or three notifications, just to ensure that receiving more than one notification works.

File: Sources/PostgreSQL/Connection/PostgreSQLConnection+NotifyAndListen.swift (hunk 3 above)
Comment (pedantix) [reply Sources/PostgreSQL/Connection/PostgreSQLConnection+NotifyAndListen.swift]: I welcome a PR to enhance this with all your helpful suggestions! Each Listen command has to be executed individually so the only modification needed should be to add "unlisten" statements. https://www.postgresql.org/docs/10/static/sql-listen.html you can read all about the technicals here. That would be a great enhanement.

File: Tests/PostgreSQLTests/PostgreSQLConnectionTests.swift (hunk 2 above)
Comment (pedantix) [reply Tests/PostgreSQLTests/PostgreSQLConnectionTests.swift]: is this part of the XCTestSuite? if so cool I am such a newb I had no idea there were these great tools for testing asynchronously I am going to look into that now.

File: Sources/PostgreSQL/Connection/PostgreSQLConnection+NotifyAndListen.swift (hunk 4 above)
Comment (pedantix) [reply Sources/PostgreSQL/Connection/PostgreSQLConnection+NotifyAndListen.swift]: https://www.postgresql.org/docs/10/static/sql-listen.html it's in the manual

File: Sources/PostgreSQL/Connection/PostgreSQLConnection.swift (hunk 1 above)
Comment (pedantix) [reply Sources/PostgreSQL/Connection/PostgreSQLConnection.swift]: easy enough

File: Tests/PostgreSQLTests/PostgreSQLConnectionTests.swift (hunk 2 above)
Comment (pedantix) [reply Tests/PostgreSQLTests/PostgreSQLConnectionTests.swift]: thanks that is such a great thing to know its going to make this Redis and other Nio based interfaces so much easier to work with

File: Sources/PostgreSQL/Connection/PostgreSQLConnection+NotifyAndListen.swift (hunk 3 above)
Comment (pedantix) [reply Sources/PostgreSQL/Connection/PostgreSQLConnection+NotifyAndListen.swift]: I guess I can make this happen it really shouldnt be so bad, its really just the cleanup thats kind of more then I expected .. .meh no big deal!

File: Tests/PostgreSQLTests/PostgreSQLConnectionTests.swift (hunk 2 above)
Comment (pedantix) [reply Tests/PostgreSQLTests/PostgreSQLConnectionTests.swift]: acctually this is a bit harder then it should be mostly because depending on the implementation and configuration you may or may not get a  channel id in the package back from the postgresql connection. this of course could be gotten around, I welcome a PR

File: Tests/PostgreSQLTests/PostgreSQLConnectionTests.swift [hunk 6]
```diff
@@ -370,7 +398,8 @@ extension PostgreSQLConnection {
     static func makeTest() throws -> PostgreSQLConnection {
         let hostname: String
         #if Xcode
-        hostname = (try? Process.execute("docker-machine", "ip")) ?? "192.168.99.100"
+        //hostname = (try? Process.execute("docker-machine", "ip")) ?? "192.168.99.100"
+        hostname = "localhost"
```
Comment (MrMage) [inline Tests/PostgreSQLTests/PostgreSQLConnectionTests.swift]: Unrelated?

File: Tests/PostgreSQLTests/PostgreSQLConnectionTests.swift [hunk 7]
```diff
@@ -342,6 +343,32 @@ class PostgreSQLConnectionTests: XCTestCase {
...
+
+        try listenConn.listen(channelName) { text in
+            if text == messageText {
+                completionHandlerExpectation1.fulfill()
+            } else if text == finalMessageText {
+                completionHandlerExpectation2.fulfill()
+            }
+        }.catch({ err in XCTFail("error \(err)") })
+
+        try notifyConn.notify(channelName, message: messageText).wait()
```
Comment (MrMage) [inline Tests/PostgreSQLTests/PostgreSQLConnectionTests.swift]: Awesome, thanks a lot!

File: Sources/PostgreSQL/Connection/PostgreSQLConnection.swift [hunk 8]
```diff
@@ -134,8 +134,16 @@ public final class PostgreSQLConnection {
         }
     }
 
+    internal var beforeClose: ((PostgreSQLConnection) -> Future<Void>)?
```
Comment (MrMage) [inline Sources/PostgreSQL/Connection/PostgreSQLConnection.swift]: In case this gets set by two different callers, some resources might not get cleaned up. How about making this an array of close handlers that you can append to? (Just a suggestion.)

File: Tests/PostgreSQLTests/PostgreSQLConnectionTests.swift (hunk 2 above)
Comment (pedantix) [reply Tests/PostgreSQLTests/PostgreSQLConnectionTests.swift]: thats just not true. I can show you screen shots of the failure where postgres only sends one null terminated string back from the server my prefered implementation would be,

```
struct PostgreSQLNotificationResponse: Decodable {
    /// The message coming from PSQL
    let channelId: String
    let message: String
    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        _ = try container.decode(Int32.self) // message length
        _ = try container.decode(Int32.self) // process id of message
        channelId = try container.decode(String.self)
        message = try container.decode(String.self)
    }
}

```

but when I run that against a psql 9 server I get the following error !
```PostgreSQLConnectionTests.swift:361: error: -[PostgreSQLTests.PostgreSQLConnectionTests testNotifyAndListen] : failed - error ⚠️ [BitsError.nullTerminatedString: This was not available in the buffer]
```

I dont get that error against other psql servers(that are 10). But I do this relatively terrible coalesce thing so depending on the message format it decodes
```
struct PostgreSQLNotificationResponse: Decodable {
    /// The message coming from PSQL
    let message: String
    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        _ = try container.decode(Int32.self) // message length
        _ = try container.decode(Int32.self) // process id of message
        let channelId = try container.decode(String.self)
        let message = try? container.decode(String.self)
        self.message = message ?? channelId
    }
}
```

File: Sources/PostgreSQL/Connection/PostgreSQLConnection+NotifyAndListen.swift (hunk 3 above)
Comment (pedantix) [reply Sources/PostgreSQL/Connection/PostgreSQLConnection+NotifyAndListen.swift]: I think that is like asking what happens if I make a query and it does not execute because of a query heppens then? If this were GRPC or AMQP I would totally agree with you that there should be a required failurehandler. but this is not a designated message broker. This seems like a good enahncement but not at this level this should be more at the top level of the connection.... which currently if there is a failure you can "catch" the error by virtue of the way eventloopfutures work. If you want something water proof it is beyond the scope of this PR as that would be an upstream change

File: Tests/PostgreSQLTests/PostgreSQLConnectionTests.swift (hunk 2 above)
Comment (MrMage) [reply Tests/PostgreSQLTests/PostgreSQLConnectionTests.swift]: Sorry, what error do you get? That code block shows up as empty to me.

Also, if Postgres randomly only returns one string when the documentation says otherwise, I think that might need more investigation or explanation. I read the documentation that the message might be omitted if the caller of "notify" didn't provide one, but the channel name should be there.

File: Sources/PostgreSQL/Connection/PostgreSQLConnection+NotifyAndListen.swift (hunk 3 above)
Comment (MrMage) [reply Sources/PostgreSQL/Connection/PostgreSQLConnection+NotifyAndListen.swift]: But how is this supposed to be used if I can't rely on getting notified, because the connection might close at any time without notice? I agree that this might not be part of this PR, but I think it should arrive least be clear what happens/what to do if the listening connection closes.

File: Tests/PostgreSQLTests/PostgreSQLConnectionTests.swift (hunk 2 above)
Comment (pedantix) [reply Tests/PostgreSQLTests/PostgreSQLConnectionTests.swift]: the code is pretty simplisitic I don't see how notify could work without a channel, I see where there is a a potential case insensitivity issue.... but  not what you are talking about

File: Sources/PostgreSQL/Connection/PostgreSQLConnection+NotifyAndListen.swift (hunk 3 above)
Comment (pedantix) [reply Sources/PostgreSQL/Connection/PostgreSQLConnection+NotifyAndListen.swift]: same question... what happens if the connection to PostgreSQL fails not if a notify/Listen handler fails. if you want to open a separate issue I am sure you myself or one of the many other members of the community can work on this. This is still Out of scope

**Comment by pedantix** [comment]
SOB mismerged it should have been merged into my branch. DOing a pr to revert

File: Tests/PostgreSQLTests/PostgreSQLConnectionTests.swift (hunk 6 above)
Comment (pedantix) [reply Tests/PostgreSQLTests/PostgreSQLConnectionTests.swift]: this is from my mismerge, waiting on @tanner0101  to berate me, I dont have the ability to roll back a merge or I would have already
